File: app/src/main/java/com/example/spring/StartSubscriptionService.java
package com.example.spring;

import com.example.spring.error.ConflictException;
import com.example.spring.error.NotFoundException;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Start a subscription for an account on a plan with a seat limit.
 */
@Service
public class StartSubscriptionService {

    private final EntitlementsRepository entitlementsRepository;
    private final ApplicationEventPublisher springEventPublisher;

    public StartSubscriptionService(EntitlementsRepository entitlementsRepository, ApplicationEventPublisher springEventPublisher) {
        this.entitlementsRepository = entitlementsRepository;
        this.springEventPublisher = springEventPublisher;
    }

    @Transactional
    public Subscription handle(StartSubscriptionRequest request) {
        // === GEN_REGION_START: service-body ===
        // spec-driven body for StartSubscription.
        // intent: Start a subscription for an account on a plan with a seat limit.
        // requires: EntitlementsRepository
        // emits events: SubscriptionStarted
        // return: Subscription (entity); spec-driven body constructs from request inputs + persists via EntitlementsRepository.saveSubscription.
        Subscription value = new Subscription(request.tenantId(), request.subscriptionId(), request.accountId(), request.planId(), request.status(), request.seatLimit(), request.currentSeats(), request.startedAt(), request.renewsAt());
        Subscription saved = this.entitlementsRepository.saveSubscription(value);
        this.springEventPublisher.publishEvent(new SubscriptionStarted(saved.accountId(), saved.planId(), saved.seatLimit(), saved.startedAt(), saved.subscriptionId(), saved.tenantId()));
        return saved;
        // === GEN_REGION_END: service-body ===
    }
}

File: app-python/entitlements_app/services/start_subscription_service.py
"""Start a subscription for an account on a plan with a seat limit."""

from decimal import Decimal
import re

from fastapi import HTTPException, status

from entitlements_app.requests.start_subscription_request import StartSubscriptionRequest
from entitlements_app.entities import Subscription
from entitlements_app.ports import EntitlementsRepository


class StartSubscriptionService:
    """Service handling Command `StartSubscription`."""

    def __init__(self, entitlements_repository: EntitlementsRepository) -> None:
        self.entitlements_repository: EntitlementsRepository = entitlements_repository

    def handle(self, request: StartSubscriptionRequest) -> Subscription:
        # spec-driven body for Command StartSubscription.
        # intent: Start a subscription for an account on a plan with a seat limit.
        # requires: EntitlementsRepository
        # emits events: SubscriptionStarted
        # return: Subscription (entity); construct from request inputs and persist via EntitlementsRepository.saveSubscription.
        value = Subscription(tenantId=request.tenantId, subscriptionId=request.subscriptionId, accountId=request.accountId, planId=request.planId, status=request.status, seatLimit=request.seatLimit, currentSeats=request.currentSeats, startedAt=request.startedAt, renewsAt=request.renewsAt)
        return self.entitlements_repository.saveSubscription(value)
